applications/schema/schema/application.schema.json

```diff
@@ -0,0 +1,61 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "YAML Application Description",
+  "description": "A full description of a YAML application AICA framework.",
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
+    "on_start": {
+      "title": "On Start",
+      "description": "on start",
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "load_controller": {
+          "$ref": "load_controller.schema.json"
+        },
+        "switch_controllers": {
+          "$ref": "switch_controllers.schema.json"
+        },
+        "load": {
+          "$ref": "load.schema.json"
+        }
+      }
+    },
+    "hardware_interfaces": {
+      "title": "Hardware Interfaces",
+      "description": "hardware interfaces",
+      "additionalProperties": false,
+      "oneOf": [
+        {
+          "type": "array",
+          "items": {
+            "$ref": "hardware_interfaces.schema.json"
+          }
+        },
+        {
+          "$ref": "hardware_interfaces.schema.json"
+        }
+      ]
```

Comment:
Well because if you just have
```
hardware_interfaces:
  papa:
    urdf: /home/ros2/examples/urdf/franka_panda.urdf
    rate: 500
    controllers:
      broadcaster:
        plugin: joint_state_broadcaster/JointStateBroadcaster
      twist_controller:
        plugin: compliant_twist_controller/CompliantTwistController
        parameters:
          linear_principle_damping: 50.0
          linear_orthogonal_damping: 50.0
          angular_stiffness: 3.0
          angular_damping: 3.0
```
this will be parsed as an object and then you get a warning that you have an object instead of an array

Reply:
Let me explain a bit more, maybe you are aware of this, maybe not.
```
hardware_interfaces:
  papa:
    urdf: 1
  quebec:
    urdf: 2
```
In the case above, the schema interprets everything below `hardware_interfaces` as an object with this syntax, whereas
```
hardware_interfaces:
  - papa:
      urdf: 1
  - quebec:
      urdf: 2
```
is interpreted as array of objects. So if we were to remove all those `oneOf`s (there are several, for the same reason) we would have to keep the object and discard the array of object, because as mentioned above, type array is not possible for our current apps, only if we would enforce the `- ` which is a bit weird for yaml files.

Therefore, I decided to support both variants with all these `oneOf` cases

Reply:
Sorry for the late followup to this review. I guess I got it the wrong way around in my original comment, but the main point is that we shouldn't support both if one way is "more correct". The schema should just allow an object of type `"$ref": "<subobject>.schema.json"` and not the array of objects. I would say that if the YAML interface can actually parse an array of interface objects with the `- papa, - quebec` notation, then it is an unintentional side-effect and not officially supported.